src/loki/reach.cc

```diff
@@ -41,38 +49,49 @@ directed_reach SimpleReach(const DirectedEdge* edge,
     if (node_filter(node))
       return;
     // otherwise we enqueue it
-    queue.insert(node_id);
+    queue_.insert(node_id);
     // and we enqueue it on the other levels
     for (const auto& transition : tile->GetNodeTransitions(node))
-      queue.insert(transition.endnode());
+      queue_.insert(transition.endnode());
     // and we remember how many duplicates we enqueued
-    transitions += node->transition_count();
+    transitions_ += node->transition_count();
   };
 
+  // TODO: here we stay extra conservative by avoiding starting on a simple restriction because we
+  // TODO: dont have predecessor information in the simple reach expansion so we bail 1 edge earlier
+  // NOTE: we can expand from the end of a complex restriction because it cant possibly be on our path
+  // and we can expand from the start of a complex restriction because only the end would mark a
+  // potential stopping point (maybe a path followed the restriction)
+
   // seed the expansion with a place to start expanding from
-  if (edge_filter(edge) > 0)
+  if (edge_filter(edge) > 0 && !edge->restrictions())
     enqueue(edge->endnode());
 
   // get outbound reach by doing a simple forward expansion until you either hit the max_reach
   // or you can no longer expand
-  while (direction & kOutbound && queue.size() + done.size() - transitions < max_reach &&
-         !queue.empty()) {
+  while (direction & kOutbound && queue_.size() + done_.size() - transitions_ < max_reach &&
+         !queue_.empty()) {
     // increase the reach and get the nodes id
-    auto node_id = GraphId(*done.insert(*queue.begin()).first);
+    auto node_id = GraphId(*done_.insert(*queue_.begin()).first);
     // pop the node from the queue
-    queue.erase(queue.begin());
+    queue_.erase(queue_.begin());
     // expand from the node
     if (!reader.GetGraphTile(node_id, tile))
       continue;
     for (const auto& edge : tile->GetDirectedEdges(node_id)) {
+      // TODO: we'd rather say !edge.end_simple_restriction() and not !edge.restrictions()
+      // TODO: but we'd need the predecessor information to do that so we punt 1 edge earlier
+      // NOTE: we can go through the start of the restriction because only the end would mark a
+      // potential stopping point (maybe a path followed the restriction)
```

Comment:
similar situation here. if we see the end of  a complex restriction or the beginning of a simple one, we must stop because our path may have been over either one of them